app/charts/shared/imputation.tsx

```diff
@@ -0,0 +1,82 @@
+import { interpolate } from "d3-interpolate";
+
+export const interpolateZerosValue = () => {
+  return Number.MIN_VALUE;
+};
+
+export const interpolateTemporalLinearValue = ({
+  previousValue,
+  nextValue,
+  previousTime,
+  currentTime,
+  nextTime,
+}: {
+  previousValue: number;
+  nextValue: number;
+  previousTime: number;
+  currentTime: number;
+  nextTime: number;
+}) => {
+  return interpolate(
+    previousValue,
+    nextValue
+  )((currentTime - previousTime) / (nextTime - previousTime));
+};
+
+interface TemporalSeriesBeforeImputationEntry {
+  date: Date;
+  value: number | null;
+}
+
+interface TemporalSeriesAfterImputationEntry {
+  date: Date;
+  value: number;
+}
+
+export const imputeTemporalLinearSeries = ({
+  dataSortedByX,
+}: {
+  dataSortedByX: Array<TemporalSeriesBeforeImputationEntry>;
+}): Array<TemporalSeriesAfterImputationEntry> => {
+  const indexedData: Array<[TemporalSeriesBeforeImputationEntry, number]> =
+    dataSortedByX.map((d, i) => [d, i]);
+  const presentDataIndexes = indexedData
+    .filter((d) => d[0].value !== null)
+    .map((d) => d[1]);
+  const missingDataIndexes = indexedData
+    .filter((d) => d[0].value === null)
+    .map((d) => d[1]);
```

Comment:
Mostly a nit but this still creates some unnecessary Arrays that you don't really need (and have to be garbage-collected).

```ts
const indexedData: Array<[TemporalSeriesBeforeImputationEntry, number]> =
    dataSortedByX.map((d, i) => [d, i]); // -> new Array!
  const presentDataIndexes = indexedData
    .filter((d) => d[0].value !== null) // -> new Array!
    .map((d) => d[1]); // -> new Array!
  const missingDataIndexes = indexedData
    .filter((d) => d[0].value === null) // -> new Array!
    .map((d) => d[1]); // -> new Array!
```

You could combine `.filter().map()` into one loop with `dataSortedByX.flatMap((d, i) => d.value !== null ? [i] : [])` but if you wanted to optimize this a bit more, a good old-fashioned for-loop is the way to go (less "functional" but arguably even more readable?). This way you only need one loop:

```ts
let presentDataIndexes = [];
let missingDataIndexes = [];
for (let i = 0; i < dataSortedByX.length; i++) {
  if (dataSortedByX[i].value !== null) {
    presentDataIndexes.push(i)
  } else {
    missingDataIndexes.push(i)
  }
}
```

Reply:
👍  for jeremy approach if this is a tight loop.

Another approach with still intermediate arrays (for the record): 

```
const { true: presentDataIndexes, false: missingDataIndexes } = mapValues(groupBy(indexedData, d => d[0].value === null), x => x[1])
```